### ISSUE TITLE
Build fails with configuration in `allprojects` section with 0.73.0

### ISSUE THREAD

**Issue by tony-schellenberg** [autonomousapps/dependency-analysis-android-gradle-plugin#412]
**Plugin version**
0.73.0

**Gradle version**
6.8.3

**Android Gradle Plugin (AGP) version**
Not an android build

**Describe the bug**
Updating to version `0.73.0` results in build issues.
It's reporting errors in configuration sections that seem to work with previous versions.

**To Reproduce**
Here's a `build.gradle` file that demonstrates the issues:
```groovy
plugins {
    id 'com.autonomousapps.dependency-analysis' version '0.73.0'
}

wrapper {
    gradleVersion = '6.8.3'
}

allprojects {
    apply plugin: "com.autonomousapps.dependency-analysis"

    // Per: https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/wiki/Customizing-plugin-behavior#failure-conditions-and-filtering

    // Adding in either of the two config blocks below breaks the build
    dependencyAnalysis {
        issues {
            onAny {
                severity('fail')
            }
        }
    }

    //dependencyAnalysis {
    //    all {
    //        issues {
    //            onAny {
    //                severity('fail')
    //            }
    //        }
    //    }
    //}
}

```
1. Copy the above `build.gradle` file to a file
2. Attempt to execute the wrapper task (i.e. `gradle wrapper` or `./path/gradlew wrapper`)
3. There is an error
4. Change the version of the plugin to `0.72.0`
5. Run the wrapper task
6. It succeeds with the earlier version

**Expected behavior**
The task should run without an error

**Additional context**
I tested with `0.71.0` and `0.72.0` and both of those seemed fine.

I included two code blocks related to the documentation link and both seem to have problems.
If both code blocks are removed, the build succeeds.

This is a very simplified example without any actual sub projects, Java code/dependencies, or other project setup but the errors that are demonstrated here are also being seen in our actual projects that are working fine with version `0.71.0` currently but are having issues with `0.73.0`.

**Comment by autonomousapps** [comment]
Thank you for the report. I suspect my documentation is out of date (and I'm working on upgrading it now). To achieve what it seems you want to, this should work:

```groovy
dependencyAnalysis {
  issues {
    // The "all" here refers to "all projects".
    all {
      onAny {
        severity('fail')
      }
    }
    // This is the alternative to "all". It configures just a single subproject, by path.
    project(":lib") {
      // as above
    }
  }
}
```
Your first example used a function that has been deprecated for a while, and which was removed in 0.73.0, in anticipation of the upcoming 1.0 release. I don't think your second example ever worked.

**Comment by autonomousapps** [comment]
Ohh, I see what's happening now. You put that in the `allprojects` block. The following is legal _only_ in subprojects
```groovy
dependencyAnalysis {
  issues {
    onAny {
      severity('fail')
    }
  }
}
```
That would fail in the root project, as the root needs the `all {}` or `project(":lib") {}` variant. If you want to configure these rules for the entire project, you're better off just putting that configuration _outside of_ `allprojects {}` and using the root extension variant of the code (from my earlier comment).

**Comment by tony-schellenberg** [comment]
I gave that a try and haven't been able to find a configuration that works for us yet. I've attached some more complete examples here.

Something like this is what we currently have today with version `0.71.0`. It uses the `allprojects` block which I understand isn't supported but it's what we've been using so far.
```groovy
plugins {
    id 'com.autonomousapps.dependency-analysis' version '0.71.0'
}

wrapper {
    gradleVersion = '6.8.3'
}

allprojects {
    apply plugin: 'java'
    apply plugin: 'checkstyle'
    apply plugin: 'com.autonomousapps.dependency-analysis'

    dependencyAnalysis {
        issues {
            onAny {
                severity('fail')
            }
        }
    }
    check.dependsOn buildHealth
}
```

When I try to use alternative configurations, I'm not able to get the `check.dependsOn` statement to work. I can pull chunks out into the root project directly and/or use subprojects but I can't seem to get it to function with the additional dependency.

Here's an example of something I tried that doesn't work:
```groovy
plugins {
    id 'com.autonomousapps.dependency-analysis' version '0.73.0'
}

wrapper {
    gradleVersion = '6.8.3'
}

apply plugin: 'java'
apply plugin: 'checkstyle'
apply plugin: 'com.autonomousapps.dependency-analysis'
dependencyAnalysis {
    issues {
        all {
            onAny {
                severity('fail')
            }
        }
    }
}
check.dependsOn buildHealth

subprojects {
    apply plugin: 'java'
    apply plugin: 'checkstyle'
    apply plugin: 'com.autonomousapps.dependency-analysis'

    check.dependsOn buildHealth
}
```

I'm going to keep working away at it but any advice or suggestions would be appreciated.

**Comment by autonomousapps** [comment]
As I understand it, you want the `check` task that exists for each project to depend on `buildHealth`. I can think of a couple of ways to achieve something like that. Here's an example

**root build.gradle**
```groovy
plugins {
  id 'com.autonomousapps.dependency-analysis' version '0.73.0'
  id 'java' // it's unusual to apply this plugin instead of java-library
  id 'checkstyle'
}

dependencyAnalysis {
    issues {
        all {
            onAny {
                severity('fail')
            }
        }
    }
}

check.dependsOn buildHealth

subprojects {
  check.dependsOn ':buildHealth'
  // or
  check.dependsOn projectHealth
  // or
  check.dependsOn aggregateAdvice
}
```
Couple of things to note here:
1. You don't need to apply this plugin in the `subprojects` block. When you apply the plugin in the root, it auto-applies itself to all subprojects.
2. `check.dependsOn ':buildHealth'` refers to the `':buildHealth'` task by path instead of with an accessor.
3. Every project has a `projectHealth` task registered, which is like `buildHealth` but for just that project.
4. Every project has an `aggregateAdvice` task registered, which is basically equivalent to `projectHealth`. The difference is that `buildHealth` literally depends on `aggregateAdvice` (you can verify by running `./gradlew buildHealth -n` and inspect the output), while `projectHealth` is more of an alias and intended as a project-level analog to `buildHealth`.

I would personally recommend having the subprojects depend on `projectHealth`, as I think that most closely matches your intention. I don't think it makes sense for `lib:check` to trigger a full build, which `buildHealth` does.

Hope that helps.

**Comment by tony-schellenberg** [comment]
I really appreciate the help and support on this. I've got something up and running with `0.73.0`.

The key piece for me seemed to be switching to `projectHealth`. I'm not sure if that was expected or not. When I copied the `build.gradle` setup from your last post and tried to run `./gradlew clean check`, it fails on setting the dependency on `buildHealth`. Switching it to `projectHealth` seems to resolve that issue.

**Comment by autonomousapps** [comment]
Glad to hear you've got something working! 

On June 15, 2021, GitHub ***@***.***> wrote:
> Closed #412 <https://github.com/autonomousapps/dependency-analysis-
> android-gradle-plugin/issues/412>.
>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/autonomousapps/dependency-analysis-android-gradle-
plugin/issues/412#event-4893810017>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/ABJG5PJFNS5EGGPAO6CGSGTTS6XIHANCNFSM46O5LJSQ>.
